databuilder/typing_utils.py

```diff
@@ -0,0 +1,143 @@
+        type_matches(dict[str, str], dict[T, T], ctx)
+
+    The first because the value of T is internally inconsistent; the second because
+    although it is internally consistent it is inconsistent with the previous example
+    whose context it shares.
+    """
+    # If `spec` is a type variable
+    if isinstance(spec, typing.TypeVar):
+        # Check the type constraints are met, if any
+        if spec.__constraints__:
+            if not any(
+                type_matches(type_, subspec, typevar_context)
+                for subspec in spec.__constraints__
+            ):
+                return False
+        # If we've already assigned a value for this variable (and it wasn't the Any
+        # type) then check it matches
+        if spec in typevar_context and typevar_context[spec] != typing.Any:
+            if typevar_context[spec] != type_:
+                return False
+        # Otherwise record this value as the value of the TypeVar
+        else:
+            typevar_context[spec] = type_
+        return True
+
+    # `Any` is the easy case: it just matches everything
+    if type_ is typing.Any or spec is typing.Any:
+        return True
+
+    spec_origin = typing.get_origin(spec)
+    spec_args = typing.get_args(spec)
+
+    # If there's no origin type that means `spec` is an ordinary class
+    if spec_origin is None:
+        return type_ is not None and issubclass(type_, spec)
+    elif spec_origin is typing.Union:
+        # For union types we just need to match one of the arguments
+        return any(type_matches(type_, arg, typevar_context) for arg in spec_args)
+    else:
+        # Otherwise we get origin and args for `type_` and check that each element
+        # matches
+        type_origin = typing.get_origin(type_)
+        if not type_matches(type_origin, spec_origin, typevar_context):
+            return False
+        type_args = typing.get_args(type_)
+        for type_arg, spec_arg in zip(type_args, spec_args):
+            if not type_matches(type_arg, spec_arg, typevar_context):
+                return False
+        return True
+
+
+@singledispatch
+def get_typespec(value):
+    """
+    Return a type specification for the supplied value. In simple cases this is just
+    like `type()` e.g.
+
+        get_typespec(1) == int
+
+    But this function knows how to destructure some container types so that e.g.
+
+        get_typespec({1, 2, 3}) == Set[int]
+
+    Additionally, through the magic of singledispatch, it can be taught to destructure
+    other container types it doesn't yet know about.
+    """
+    return type(value)
+
+
+@get_typespec.register(Set)
+def get_typespec_for_set(value):
+    member_types = {get_typespec(i) for i in value}
+    if len(member_types) > 1:
+        raise TypeError(f"Sets must be of homogeneous type: {value!r}")
+    elif member_types:
+        member_type = list(member_types)[0]
+    else:
+        # Allow empty sets
+        member_type = typing.Any
```

Comment:
I'm slightly nervous about letting `Any` creep in here (and in `Mapping`) given the shortcut effect it has elsewhere in the implementation (in `type_matches()`). Would `object` be appropriate instead?

Reply:
`object` doesn't give us the desired behaviour here because we want empty sets to match `set[int]` which `set[Any]` does but `set[object]` doesn't. I think this is one the cases where the nasty corners of our approach (converting values to typespecs rather than checking the values directly) poke out.

Reply:
Why doesn't `set[object]` match? Lack of covariance?

Reply:
Yes, I think that's the reason. As we're currently doing things `set[int]` matches `set[object]` but not the reverse. "Satisfies" might be a better word here than "matches" as it doesn't imply symmetry.

Reply:
Does my concern in the first part of my comment make sense?

Reply:
Not entirely, in that `Any` only shortcuts at the point we reach it; it's not like the mere presence of `Any` on either side of the type check makes the checker immediately give up. So `set[Any]` will satisfy `set[ ... <whatever> ...]`, but not `dict[Any]`. So I think we get the right behaviour here.